apm-agent-core/src/main/java/co/elastic/apm/bci/bytebuddy/SoftlyReferencingTypePoolCache.java

```diff
@@ -0,0 +1,110 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.bci.bytebuddy;
+
+import co.elastic.apm.util.ExecutorUtils;
+import com.blogspot.mydailyjava.weaklockfree.WeakConcurrentMap;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.pool.TypePool;
+
+import java.lang.ref.SoftReference;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Caches {@link TypeDescription}s which speeds up type matching -
+ * especially when the matching requires lookup of other {@link TypeDescription}s.
+ * Such as when in order to match a type, it's superclass has to be determined.
+ * Without a type pool cache those types would have to be re-loaded from the file system if their {@link Class} has not been loaded yet.
+ * <p>
+ * In order to avoid {@link OutOfMemoryError}s because of this cache,
+ * the {@link TypePool.CacheProvider}s are wrapped in a {@link SoftReference}
+ * </p>
+ */
+public class SoftlyReferencingTypePoolCache extends AgentBuilder.PoolStrategy.WithTypePoolCache {
+
+    /*
+     * Weakly referencing ClassLoaders to avoid class loader leaks
+     * Softly referencing the type pool cache so that it does not cause OOMEs
+     */
+    private final WeakConcurrentMap<ClassLoader, SoftReference<TypePool.CacheProvider>> cacheProviders =
+        new WeakConcurrentMap<ClassLoader, SoftReference<TypePool.CacheProvider>>(false);
+    private final AtomicLong lastAccess = new AtomicLong(System.currentTimeMillis());
+    private final ElementMatcher<ClassLoader> ignoredClassLoaders;
+
+    public SoftlyReferencingTypePoolCache(final TypePool.Default.ReaderMode readerMode,
+                                          final int clearIfNotAccessedSinceMinutes, ElementMatcher.Junction<ClassLoader> ignoredClassLoaders) {
+        super(readerMode);
+        ExecutorUtils.createSingleThreadSchedulingDeamonPool("type-cache-pool-cleaner", 1)
+            .scheduleWithFixedDelay(new Runnable() {
+                @Override
+                public void run() {
+                    clearIfNotAccessedSince(clearIfNotAccessedSinceMinutes);
```

Comment:
I believe that if you change to identify stale cache providers individually (comment above), you don't need to use a weak map, which for big maps can make significant difference in performance